test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java

```diff
@@ -0,0 +1,562 @@
+import java.util.regex.Pattern;
+
+/**
+ * Parse the hotspot pid file of the test VM to match all @IR rules.
+ */
+class IRMatcher {
+    private static final boolean PRINT_IR_ENCODING = Boolean.parseBoolean(System.getProperty("PrintIREncoding", "false"));
+    private final Map<String, IRMethod> compilations;
+    private final Class<?> testClass;
+    private final Map<Method, List<String>> fails;
+    private final String hotspotPidFileName;
+    private IRMethod irMethod; // Current IR method to which rules are applied
+    private Method method; // Current method to which rules are applied
+    private IR irAnno; // Current IR annotation that is processed.
+    private int irRuleIndex; // Current IR rule index;
+
+    public IRMatcher(String hotspotPidFileName, String irEncoding, Class<?> testClass) {
+        this.compilations =  new LinkedHashMap<>();
+        this.fails = new HashMap<>();
+        this.testClass = testClass;
+        this.hotspotPidFileName = hotspotPidFileName;
+        setupTestMethods(irEncoding);
+        if (TestFramework.VERBOSE || PRINT_IR_ENCODING) {
+            System.out.println("Read IR encoding from test VM:");
+            System.out.println(irEncoding);
+        }
+        parseHotspotPidFile();
+        applyRules();
+    }
+
+    /**
+     * Sets up a map testname -> IRMethod (containing the PrintIdeal and PrintOptoAssembly output for testname).
+     */
+    private void setupTestMethods(String irEncoding) {
+        Map<String, Integer[]> irRulesMap = parseIREncoding(irEncoding);
+        for (Method m : testClass.getDeclaredMethods()) {
+            method = m;
+            IR[] irAnnos =  m.getAnnotationsByType(IR.class);
+            if (irAnnos.length > 0) {
+                // Validation of legal @IR attributes and placement of the annotation was already done in Test VM.
+                Integer[] ids = irRulesMap.get(m.getName());
+                TestFramework.check(ids != null, "Should find method name in validIrRulesMap for " + m);
+                TestFramework.check(ids.length > 0, "Did not find any rule indices for " + m);
+                TestFramework.check(ids[ids.length - 1] < irAnnos.length, "Invalid IR rule index found in validIrRulesMap for " + m);
+                if (ids[0] != IREncodingPrinter.NO_RULE_APPLIED) {
+                    // If -1, than there was no matching IR rule for the given conditions.
+                    compilations.put(m.getName(), new IRMethod(m, ids, irAnnos));
+                }
+            }
+        }
+    }
+
+    /**
+     * Read the IR encoding emitted by the test VM to decide if an @IR rule must be checked for a method.
+     */
+    private Map<String, Integer[]>  parseIREncoding(String irEncoding) {
+        Map<String, Integer[]> irRulesMap = new HashMap<>();
+        String patternString = "(?<=" + IREncodingPrinter.START + "\\R)[\\s\\S]*(?=" + IREncodingPrinter.END + ")";
+        Pattern pattern = Pattern.compile(patternString);
```

Comment:
`patternString` is effectively a constant here, but you will compile it into `j.u.Pattern` on each invocation of `parseIREncoding`, it might be a better idea to introduce a static field that holds a precompiled patter.